apps/meteor/server/services/video-conference/service.ts

```diff
@@ -0,0 +1,238 @@
+		const call = await this.VideoConference.findOneById(callId);
+		if (!call || !isDirectVideoConference(call)) {
+			throw new Error('invalid-call');
+		}
+
+		if (call.status !== VideoConferenceStatus.CALLING || call.endedBy || call.endedAt) {
+			throw new Error('invalid-call-status');
+		}
+
+		const user = await this.Users.findOneById(uid);
+		if (!user) {
+			throw new Error('failed-to-load-own-data');
+		}
+
+		if (call.messages.started) {
+			await this.changeMessageType(call.messages.started, 'video-direct-missed');
+		}
+
+		this.VideoConference.setEndedById(call._id, { _id: user._id, name: user.name, username: user.username });
+	}
+
+	public async get(callId: IVideoConference['_id']): Promise<IVideoConference | null> {
+		return this.VideoConference.findOneById(callId);
+	}
+
+	private async createMessage(
+		type: MessageTypesValues,
+		rid: IRoom['_id'],
+		user: IUser,
+		extraData: Partial<IMessage> = {},
+	): Promise<IMessage['_id']> {
+		const readReceipt = settings.get('Message_Read_Receipt_Enabled');
+
+		const record = {
+			t: type,
+			rid,
+			ts: new Date(),
+			u: {
+				_id: user._id,
+				username: user.username,
+			},
+			groupable: false,
+			...(readReceipt ? { unread: true } : {}),
+			...extraData,
+		};
+
+		const id = (await this.Messages.insertOne(record)).insertedId;
+		this.Rooms.incMsgCountById(rid, 1);
+		return id;
+	}
+
+	private async createDirectCallMessage(rid: IRoom['_id'], user: IUser): Promise<IMessage['_id']> {
+		return this.createMessage('video-direct-calling', rid, user);
+	}
+
+	private async changeMessageType(messageId: string, newType: MessageTypesValues): Promise<void> {
+		this.Messages.setTypeById(messageId, newType);
+	}
+
+	private async startDirect(caller: IUser['_id'], { _id: rid, uids }: AtLeast<IRoom, '_id' | 'uids'>): Promise<DirectCallInstructions> {
+		const callee = uids?.filter((uid) => uid !== caller).pop();
+		if (!callee) {
+			// Are you trying to call yourself?
+			throw new Error('invalid-call-target');
+		}
+
+		const user = await this.Users.findOneById<IUser>(caller, {});
+		if (!user) {
+			throw new Error('failed-to-load-own-data');
+		}
+
+		const callId = await this.VideoConference.createDirect(rid, {
+			_id: user._id,
+			name: user.name,
+			username: user.username,
+		});
+
+		const url = await this.generateNewUrl(callId);
+		this.VideoConference.setUrlById(callId, url);
+
+		const messageId = await this.createDirectCallMessage(rid, user);
+		this.VideoConference.setMessageById(callId, 'started', messageId);
+
+		return {
+			type: 'direct',
+			callId,
+			callee,
+		};
+	}
+
+	private async joinDirect(
+		call: IVideoConference,
+		user: AtLeast<IUser, '_id' | 'username' | 'name'>,
+		options: VideoConferenceJoinOptions,
+	): Promise<string> {
+		const url = this.getUrl(call, user, options);
+
+		if (!call.users.find(({ _id }) => _id === user._id)) {
+			this.addUserToCall(call._id, user);
```

Comment:
should this await to make sure the user was added?
```suggestion
			await this.addUserToCall(call._id, user);
```